url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/utility/security/Gfunction.java

```diff
@@ -0,0 +1,636 @@
+		}
+	}
+	
+	public String ClientComputed(String[] ServerReturned, long[] x) throws Exception{
+		String str;
+		int len = ServerReturned.length;
+		int l = Integer.parseInt(ServerReturned[0]);
+		int r = Integer.parseInt(ServerReturned[1]);
+		if(l == -1){
+			m = r - 1;
+			String[] strs = new String[m + 1];
+			//System.out.println("==========");
+			for(int i = 0; i < m + 1; i++){
+				strs[i] = HashXtimes(ServerReturned[2 + i + 1], x[i]);
+				//System.out.println(strs[i]);
+			}
+			str = SecurityUtility.computeGeneralHashValue(strs);
+			//System.out.println("Computed : " + str);
+			str = SecurityUtility.computeGeneralHashValue(new String[]{ServerReturned[2], str});
+			String ans = rebuildVO(2 + 1 + m + 1, len - 1, ServerReturned, str);
+			return ans;
+		}else if(r == -1){
+			m = l - 1;
+			String[] strs = new String[m + 1];
+			for(int i = 0; i < m + 1; i++){
+				strs[i] = HashXtimes(ServerReturned[2 + i], x[i]);
+			}
+			str = SecurityUtility.computeGeneralHashValue(strs);
+			str = SecurityUtility.computeGeneralHashValue(new String[]{str, ServerReturned[2 + 1 + m]});
+			String ans = rebuildVO(2 + 1 + m + 1, len - 1, ServerReturned, str);
+			return ans;
+
+		}else{
+			throw new Exception("Client verified error!");
+		}
+	}
+	
+	
+	/**
+	 * It seems if isL = true, x <= val can pass; if isR = true, x >= val can pass.
+	 * 
+	 * 
+	 * */
+	public static void main(String[] args) throws Exception{
+		
+		/**
+		 * test for g function
+		 * */
+		long val = 20608, x = 20658, times = 1000;
+		ThreadMXBean bean = ManagementFactory.getThreadMXBean();
+		long start = bean.getCurrentThreadCpuTime(), cputime;
+		Gfunction gf = new Gfunction(val, 2);
+		String[]  ServerReturned = null;
+		start = bean.getCurrentThreadCpuTime();
+		for(int i = 0 ; i < times; i ++){
+			ServerReturned = gf.GenerateVeryfyPart(x, false);
+		}
+		cputime = bean.getCurrentThreadCpuTime() - start;
+		System.out.println("SP consume:\t" + cputime / times / 1000000.0 + "ms");
+		start = bean.getCurrentThreadCpuTime();
+		for(int i = 0 ; i < times; i ++){
+			if(gf.getDigest().equals(gf.ClientComputed(ServerReturned, gf.U - x)) == true){
+				//System.out.println("Pass verification!");
+			}else {
+				System.out.println("Fail verification!");
+				break;
+			}
+		}
+		cputime = bean.getCurrentThreadCpuTime() - start;
+		System.out.println("Client consume:\t" + cputime / times / 1000000.0 + "ms");
+		val = 15169; 
+		x = 15104;
+		gf = new Gfunction(val, 2);
+		ServerReturned = gf.GenerateVeryfyPart(x, true);
+		start = System.currentTimeMillis();
+		for(int i = 0 ; i < times; i ++){
+			if(gf.getDigest().equals(gf.ClientComputed(ServerReturned, x - gf.L)) == true){
+				//System.out.println("Pass verification!");
+			}else {
+				System.out.println("Fail verification!");
+				break;
+			}
+		}
+		
+		/**
+		 * test file in and file out
+		 * */
+		DataOutputStream dos = new DataOutputStream(new FileOutputStream(new File("./tmp/gf")));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *dos*. The resource is referenced at lines 626, 627. The resource is closed at line 628. There are other execution paths that do not contain closure statements, for example, when Gfunction.writeToFile throws an exception. Ensure *dos* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).